src/scala/com/github/johnynek/bazel_deps/Normalizer.scala

```diff
@@ -126,28 +134,34 @@ object Normalizer {
       }
     }
 
-  private def rewrite(g: Graph[MavenCoordinate, Unit], t: Table) = {
+  private def rewrite(g: Graph[MavenCoordinate, Unit], roots: List[MavenCoordinate], t: Table): Option[Graph[MavenCoordinate, Unit]] = {
     if (t.forall { case (_, vs) => vs.forall(_._2.isRight) }) {
-      val versions = g.nodes.groupBy(_.unversioned).mapValues(_.map(_.version))
       val canonicals = compact(t)
 
-      def retarget(g0: Graph[MavenCoordinate, Unit], n0: MavenCoordinate): Graph[MavenCoordinate, Unit] =
-        if (isKeeper(n0, t)) {
-          val dependencies = g0.hasSource(n0).map(_.destination)
-          dependencies.foldLeft(g0)(retarget)
-        }
-        else {
-          // rewrite all the edges that terminate at us
-          // things that depend on us need to start depending on someone else
-          val u = n0.unversioned
-          val version = canonicals(u)
-          val newTarget = MavenCoordinate(u, version)
-          g0.hasDestination(n0).foldLeft(g0) { (g, edge) =>
-            g.removeEdge(edge).addEdge(Edge(edge.source, newTarget, ()))
-          }
+      @annotation.tailrec
+      def addReachable(acc: Graph[MavenCoordinate, Unit], toProcess: List[UnversionedCoordinate], processed: Set[UnversionedCoordinate]): Graph[MavenCoordinate, Unit] =
+        toProcess match {
+          case Nil => acc
+          case h :: tail if processed(h) => addReachable(acc, tail, processed)
+          case h :: tail =>
+            val versionedH = MavenCoordinate(h, canonicals(h))
+            // we want to add this, and all the nodes this versionedH pointed to in original graph
+            val dependenciesUv: Set[UnversionedCoordinate] =
+              g.hasSource(versionedH).map(_.destination.unversioned)
+
+            val dependencies = dependenciesUv.map { uv =>
+              MavenCoordinate(uv, canonicals(uv))
+            }
+            // add the edges from versionedH -> deps
```

Comment:
Here's a patch that makes the test fail: https://gist.github.com/kevingessner/ce8981aecb64dd2885d19e4a35fa366a